File: ga/gpt3_5_result_2024_6/result/1968D_gpt/15_1968D.c
fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: i32 = input.trim().parse().unwrap();

    for _ in 0..t {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut values = input.trim().split_whitespace().map(|x| x.parse::<i32>().unwrap());
        let n = values.next().unwrap();
        let k = values.next().unwrap();
        let mut a = values.next().unwrap();
        let mut b = values.next().unwrap();

        let m = n.min(k);
        let mut arr = Vec::with_capacity(n as usize);
        let mut aa = Vec::with_capacity(n as usize);

        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        arr.extend(input.trim().split_whitespace().map(|x| x.parse::<i32>().unwrap()));
        
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        aa.extend(input.trim().split_whitespace().map(|x| x.parse::<i32>().unwrap()));
        
        let (mut p1, mut p2, mut p1_, mut p2_) = (0, 0, 0, 0);
        let (mut i, mut i_, mut j, mut j_) = (-1, -1, -1, -1);
        
        for x in 1..=m {
            p1 = p1.max((aa[a as usize - 1] * (k - x + 1)) as i64 + p1_);
            p1_ += aa[a as usize - 1];
            a = arr[a as usize - 1];

            p2 = p2.max((aa[b as usize - 1] * (k - x + 1)) as i64 + p2_);
            p2_ += aa[b as usize - 1];
            b = arr[b as usize - 1];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}

File: ga/glm4_result_2024_6/result/1968D_cpp_glm/15_1968D.rs
use std::io;

fn calculate(p: &Vec<usize>, a: &Vec<usize>, pos: usize, k: usize) -> u64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    
    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = std::cmp::max(mx, cur + (k as u64) * (a[pos] as u64));
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }
    
    mx
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");
    let t: usize = input.trim().parse().expect("Failed to parse t");

    for _ in 0..t {
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let mut iter = input.trim().split_whitespace();
        let n: usize = iter.next().unwrap().parse().expect("Failed to parse n");
        let k: usize = iter.next().unwrap().parse().expect("Failed to parse k");
        let ps1: usize = iter.next().unwrap().parse().expect("Failed to parse ps1");
        let ps2: usize = iter.next().unwrap().parse().expect("Failed to parse ps2");
        
        let mut p = Vec::with_capacity(n);
        let mut a = Vec::with_capacity(n);
        
        for _ in 0..n {
            p.push(iter.next().unwrap().parse().expect("Failed to parse p[i]"));
        }
        
        for _ in 0..n {
            a.push(iter.next().unwrap().parse().expect("Failed to parse a[i]"));
        }
        
        for i in 0..n {
            p[i] -= 1;
        }
        
        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);
        
        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}